---
title: "Creating an OTU table"
author: "Aldas Zarnauskas"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

```{r libraries}
library(tidyverse)
library(phyloseq)
library(ggplot2)
library(vegan)
library(ComplexHeatmap)
library(DESeq2)

theme_set(theme_classic())
```

```{r OTU and taxonomy}
#Load the OTU table
OTU <- read_delim("~/bioinformatics/microbiome_project/Microbiome_project_downstream_analysis/data/feature-table.tsv", col_names = T, delim = "\t", skip = 1)
#Make OTU IDs as rownames
OTU <- column_to_rownames(OTU, var = "#OTU ID")

#Import taxonomy table. Taxonomy table specifies taxonomic classification for each OTU ID
taxonomy_table <- read_delim("~/bioinformatics/microbiome_project/Microbiome_project_downstream_analysis/data/taxonomy.tsv", col_names = T, delim = "\t")
#Divide Taxon column into taxa. All taxon in the Taxon column is merged.
tax.clean <- separate(taxonomy_table, col = "Taxon", sep = ";", into = c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species"), remove = T, fill = "warn")
tax.clean <- apply(tax.clean, 2, function(x) gsub("^d__|^p__|^c__|^o__|^f__|^g__|^s__", "", x ))

#Substitute NA or __ cells with an empty string ""
tax.clean[is.na(tax.clean)] <- ""
tax.clean[tax.clean=="__"] <- ""

#Make Feature IDs column as rownames
tax.clean <- tax.clean %>% as.data.frame() %>%  column_to_rownames(var = "Feature ID")

#Substitute any unassigned taxon classification with unclassified + lowest assigned taxonomic classification
for (i in 1:nrow(tax.clean)){
  if (tax.clean[i,7] != ""){
    tax.clean$Species[i] <- paste(tax.clean$Genus[i], tax.clean$Species[i], sep = " ")
  } else if (tax.clean[i,2] == ""){
    kingdom <- paste("Unclassified", tax.clean[i,1], sep = " ")
    tax.clean[i, 2:7] <- kingdom
  } else if (tax.clean[i,3] == ""){
    phylum <- paste("Unclassified", tax.clean[i,2], sep = " ")
    tax.clean[i, 3:7] <- phylum
  } else if (tax.clean[i,4] == ""){
    class <- paste("Unclassified", tax.clean[i,3], sep = " ")
    tax.clean[i, 4:7] <- class
  } else if (tax.clean[i,5] == ""){
    order <- paste("Unclassified", tax.clean[i,4], sep = " ")
    tax.clean[i, 5:7] <- order
  } else if (tax.clean[i,6] == ""){
    family <- paste("Unclassified", tax.clean[i,5], sep = " ")
    tax.clean[i, 6:7] <- family
  } else if (tax.clean[i,7] == ""){
    tax.clean$Species[i] <- paste("Unclassified ",tax.clean$Genus[i], sep = " ")
  }
}
```

```{r metadata}
metadata <- read_tsv("/home/aldas/bioinformatics/microbiome_project/Microbiome_project_downstream_analysis/data/filereport_read_run_PRJNA560950_tsv.txt", col_names = T)

metadata <- metadata[, c("sample_alias", "run_accession")]

metadata.new <- data.frame(
  sample_alias = c(sort(metadata$sample_alias)),
  treatment = c(rep("Control", 5),
                rep("Tapioca", 5),
                rep("Potato", 5),
                rep("Potato", 5),
                rep("Control", 5),
                rep("Control", 5),
                rep("Maize", 5),
                rep("Tapioca", 5),
                rep("Maize", 5),
                rep("Maize", 5),
                rep("Tapioca", 5),
                rep("Tapioca", 5),
                rep("Tapioca", 5),
                rep("Potato", 5),
                rep("Maize", 5),
                rep("Control", 5),
                rep("Control", 5),
                rep("Maize", 5),
                rep("Potato", 5),
                rep("Potato", 5),
                
                rep("Potato", 5),
                rep("Maize", 5),
                rep("Tapioca", 5),
                rep("Control", 5),
                rep("Control", 5),
                rep("Tapioca", 5),
                rep("Potato", 5),
                rep("Tapioca", 5),
                rep("Maize", 5),
                rep("Maize", 5),
                rep("Control", 5),
                rep("Potato", 5),
                rep("Maize", 5),
                rep("Tapioca", 5),
                rep("Potato", 5),
                rep("Control", 5),
                rep("Control", 5),
                rep("Maize", 5),
                rep("Potato", 5),
                rep("Tapioca", 5)
                
))

metadata.new$time <- lapply(metadata.new$sample_alias, function(x) substr(x, 4, 5))
metadata.new$sampleID <- lapply(metadata.new$sample_alias, function(x) substr(x, 1, 3))
metadata.new <- metadata.new %>%  
  dplyr::mutate(dose = case_when(time == "W0" ~ 0,
                                 time == "W1" ~ 10,
                                 time == "W2" ~ 20,
                                 time == "W3" ~ 35,
                                 time == "W4" ~ 50))
metadata.merged <- merge(metadata, metadata.new, by = "sample_alias")
metadata.final <- column_to_rownames(metadata.merged, var = "run_accession")
```

```{r tree}
phylotree <- read_tree("/home/aldas/bioinformatics/microbiome_project/exported-phylogeny-align-to-tree-mafft-fasttree/tree.nwk")
```

```{r phyloseq}
OTU <- otu_table(OTU, taxa_are_rows=T)
TAX <- tax_table(tax.clean[,1:7] %>% as.matrix())
metadata <- sample_data(metadata.final)
tree <- phy_tree(phylotree)
physeq <- phyloseq(OTU, TAX, metadata, tree)
```

```{r removing}
rm(list = setdiff(ls(all.names = T), "physeq"))
```


# Data Analysis and Visualisation

```{r alpha diversity}
#In the documentation, authors state that we should use not-trimmed and unnormalised counts to calculate alpha diversity.

alpha_meas = c("Observed"
               , "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson"
               )
p1 <- plot_richness(physeq, "dose",  measures=alpha_meas)

p1$data$dose <- as.factor(p1$data$dose)

p1_observed <- as.data.frame(p1$data) %>% subset(variable == "Observed")
p1_Chao1 <- as.data.frame(p1$data) %>% subset(variable == "Chao1")
p1_ACE <- as.data.frame(p1$data) %>% subset(variable == "ACE")
p1_Shannon <- as.data.frame(p1$data) %>% subset(variable == "Shannon")
p1_Simpson <- as.data.frame(p1$data) %>% subset(variable == "Simpson")
p1_InvSimpson <- as.data.frame(p1$data) %>% subset(variable == "InvSimpson")


p1_observed.plot <- ggplot(p1_observed, aes(x=dose, y=value, fill = treatment)) +  
  geom_violin() +
  facet_grid(~ treatment) +
  labs(x="Dose grams", y="Observed alpha diversity")


p1_shannon.plot <- ggplot(p1_Shannon, aes(x=dose, y=value, fill = treatment)) +  
  geom_violin() +
  facet_grid(~ treatment) +
  labs(x="Dose grams", y="Shannon alpha diversity")
```

```{r estimate richness}
ps_control <- subset_samples(physeq, treatment=="Control")
rich_control = estimate_richness(ps_control, measures = c("Observed", "Shannon"))
wilcox.shannonC <- pairwise.wilcox.test(rich_control$Shannon, 
                                        sample_data(ps_control)$dose, 
                                        p.adjust.method = "BH")

ps_Tapioca <- subset_samples(physeq, treatment=="Tapioca")
rich_Tapioca = estimate_richness(ps_Tapioca, measures = c("Observed", "Shannon"))
wilcox.shannonT <- pairwise.wilcox.test(rich_Tapioca$Shannon, 
                                        sample_data(ps_Tapioca)$dose, 
                                        p.adjust.method = "BH")

ps_Maize <- subset_samples(physeq, treatment=="Maize")
rich_Maize = estimate_richness(ps_Maize, measures = c("Observed", "Shannon"))
wilcox.shannonM <- pairwise.wilcox.test(rich_Maize$Shannon, 
                                        sample_data(ps_Maize)$dose, 
                                        p.adjust.method = "BH")

ps_Potato <- subset_samples(physeq, treatment=="Potato")

rich_Potato = estimate_richness(ps_Potato, measures = c("Observed", "Shannon"))
wilcox.shannonP <- pairwise.wilcox.test(rich_Potato$Shannon, 
                                        sample_data(ps_Potato)$dose, 
                                        p.adjust.method = "BH")
```


```{r Bray-curtis dissimilarity}
#bray_curtis_dist <- as.matrix(vegdist(t(otu_table(physeq)), method = "bray"))
ord.meas <- ordinate(physeq, method = "PCoA", distance = "bray")
plot_ordination(physeq, ord.meas)+ geom_point(aes(colour=treatment),size=2) + 
  theme_classic() + 
  theme(text = element_text(size=18), axis.text = element_text(size=16), 
        legend.position = "right") +
  ggtitle("PCoA: Bray-Curtis")
```

```{r functions}
dta <- function(ps, treatment, taxa) {
  
  ps_object <- subset_samples(ps, treatment == eval(parse(text = "supplement")))
  sample_data(ps_object)$dose <- as.factor(sample_data(ps_object)$dose)
  ps_object <- tax_glom(ps_object, taxrank = taxa)
  
  ds <- phyloseq_to_deseq2(ps_object, ~ dose)
  rownames(ds) <- tax_table(ps_object)[, taxa] %>% as.vector()
  ds <- DESeq(ds)
  
  r1 <- results(ds, contrast = c("dose", 0, 10))
  r2 <- results(ds, contrast = c("dose", 0, 20))
  r3 <- results(ds, contrast = c("dose", 0, 35))
  r4 <- results(ds, contrast = c("dose", 0, 50))
  
  res <- list(r1 = r1, r2 = r2, r3 = r3, r4 = r4)
  return(res)
}

# log2FC_val function
log2FC_val <- function(ps, treatment, taxa) {
  # Get results from dta function
  dta_results <- dta(ps, treatment, taxa)
  
  # Define doses
  doses <- c("10", "20", "35", "50")
  
  # Initialize merged_df as NULL
  merged_df <- NULL
  
  # Iterate over doses and results
  for (i in seq_along(doses)) {
    dose <- doses[i]
    result_df <- dta_results[[i]] %>% 
      as.data.frame() %>% 
      rownames_to_column(var = taxa) %>% 
      select(taxa, log2FoldChange) %>% 
      rename_with(~ paste(dose, "g-0g", sep = ""), .cols = log2FoldChange)
    
    # Merge with the accumulated data frame
    if (is.null(merged_df)) {
      merged_df <- result_df
    } else {
      merged_df <- merge(merged_df, result_df, by = taxa, all = TRUE)
    }
  }
  
  # Convert back to a data frame with row names
  merged_df <- column_to_rownames(merged_df, var = taxa)
  return(merged_df)
}
```


```{r Heatmap}
heatmap_matrix <- function(phyla, phyla_list){
  supplements <- c("Tapioca", "Maize", "Potato", "Control")
  log2FC_results <- list()
  
  for (supplement in supplements) {
    # Compute log2FC values and extract the "Firmicute" row
    log2FC_results[[supplement]] <- log2FC_val(physeq, supplement, phyla) %>% .[phyla_list, ]
  }
  
  tapioca_maize <- merge(log2FC_results[["Tapioca"]], log2FC_results[["Maize"]], 
                         by = "row.names", suffixes = c("tapioca", "maize"))
  potato_contro <- merge(log2FC_results[["Potato"]], log2FC_results[["Control"]], 
                         by = "row.names", suffixes = c("potato", "control"))
  phyla_df <- merge(tapioca_maize, potato_contro, by = "Row.names") %>% 
    column_to_rownames(var = "Row.names") %>% as.matrix()
  return(phyla_df)
}

phylum_df <- heatmap_matrix("Phylum", c("Firmicutes"))
family_df <- heatmap_matrix("Family", c("Porphyromonadaceae","Prevotellaceae",
                                        "Bacteroidaceae", "Enterobacteriaceae"))
genera_df <- heatmap_matrix("Genus", c("Parabacteroides", "Prevotella",
                                        "Butyricimonas", "Unclassified Ruminococcaceae",
                                        "Ruminococcus", "Butyricicoccus",
                                        "Paraprevotella", "Bacteroides", "Bilophila"))


heatmap_colnames <- c(rep(c("10g-0g", "20g-0g", "35g-0g", "50g-0g"), 4))

phylum_heatmap <- Heatmap(phylum_df, 
        column_split = c(rep("Tapioca RS4", 4), rep("Maize RS4", 4), 
                         rep("Potato RS4", 4), rep("Corn Starch", 4)), 
        column_order =1:16, column_gap = unit(5, "mm"),
        show_column_names = F,
        show_row_dend = F,
        heatmap_height = unit(1, "cm"),
        col = colorRampPalette(c("darkgreen", "white", "darkgoldenrod"))(50)
        )
family_heatmap <- Heatmap(family_df, 
                          column_split = c(rep("Tapioca RS4", 4), rep("Maize RS4", 4), 
                                           rep("Potato RS4", 4), rep("Corn", 4)), 
                          column_order =1:16, column_gap = unit(5, "mm"),
                          show_column_names = F,
                          show_row_dend = F,
                          heatmap_height = unit(1.5, "cm"),
                          col = colorRampPalette(c("darkgreen", "white", "darkgoldenrod"))(50)
)
genera_heatmap <- Heatmap(genera_df, 
                          column_split = c(rep("Tapioca RS4", 4), rep("Maize RS4", 4), 
                                           rep("Potato RS4", 4), rep("Corn", 4)), 
                          column_order =1:16, column_gap = unit(5, "mm"),
                          column_labels = heatmap_colnames,
                          show_row_dend = F,
                          heatmap_height = unit(5, "cm"),
                          col = colorRampPalette(c("darkgreen", "white", "darkgoldenrod"))(50)
)

phylum_heatmap %v% family_heatmap %v% genera_heatmap
# This heatmap has four identical heatmaps for Corn Starch, Maize RS4, Potato RS4 and Tapioca RS4. This is happened because there is an issue to subset phyloseq object to contain only samples of one of this supplement. The phyloseq package function subset_samples does not work well within user defined functions. This requires further optimisation to construct a correct heatmap shown below.

```





